File: problemset/Minimum_Path_Sum.cpp
class Solution {
public:
    inline static const int INF = 0x3f3f3f3f;
    int minPathSum(vector<vector<int>>& grid) {
        int n = grid.size();
        if (!n) return 0;
        int m = grid[0].size();
        if (!m) return 0;
        vector<vector<int>> dp(n, vector<int>(m, INF));
        dp[n-1][m-1] = grid[n-1][m-1];
        for (int i = m-2; i >= 0; --i) 
            dp[n-1][i] = grid[n-1][i] + dp[n-1][i+1];
        for (int i = n-2; i >= 0; --i) 
            for (int j = m-1; j >= 0; --j) {
                int min_val = dp[i+1][j];
                if (j != m-1 && dp[i][j+1] < min_val) min_val = dp[i][j+1];
                dp[i][j] = grid[i][j] + min_val;
            }
        return dp[0][0];
    }
};

File: activity/20200904/scau-flyingbird/MinimumPathSum.go
func minPathSum(grid [][]int) int {
	if len(grid) == 0 {
		return 0
	}
	rows, cols := len(grid), len(grid[0])
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if i == 0 && j == 0 {
				continue
			} else if i == 0 {
				grid[0][j] = grid[0][j-1] + grid[0][j]
			} else if j == 0 {
				grid[i][j] = grid[i-1][j] + grid[i][j]
			} else {
				grid[i][j] = Min(grid[i-1][j], grid[i][j-1]) + grid[i][j]
			}
		}
	}
	return grid[rows-1][cols-1]
}

func Min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}
